run/markdown-preview/editor/main.js

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const express = require('express');
+const handlebars = require('handlebars');
+const { readFileSync } = require('fs');
+const renderRequest = require('./render.js');
+
+const app = express();
+app.use(express.json());
+
+let url, isAuthenticated, markdownDefault, compiledTemplate, template;
+
+const init = () => {
+  url = process.env.EDITOR_UPSTREAM_RENDER_URL;
+  if (!url) throw Error ("No configuration for upstream render service: add EDITOR_UPSTREAM_RENDER_URL environment variable");
+  isAuthenticated = !process.env.EDITOR_UPSTREAM_UNAUTHENTICATED;
+  if (!isAuthenticated) console.log("Editor: starting in unauthenticated upstream mode");
+  return {url, isAuthenticated};
```

Comment:
The main reason I've been using the pattern of `EDITOR_UPSTREAM_UNAUTHENTICATED` is that the code using the metadata server only works on Google Cloud compute platforms. However, with the new ID token helper methods rolling out, we could use the library to 1) Check if we have credentials/identity and 2) use the helper methods to mint a token locally. The upstream service can choose whether or not to verify.

If we take that approach, we should remove the `EDITOR_UPSTREAM_UNAUTHENTICATED` environment variable, and at that point I think it's worth reconsidering having this init() method and passing around a service object. Instead, it might be worth taking the approach of checking for `process.env.EDITOR_UPSTREAM_RENDER_URL` in global context and simply accessing it from process.env when the URL is needed.

This has at least some ramifications for all language variants of the sample, so raising this here and we'll update the design doc with conclusions.

@averikitsch @dinagraves @kelsk

Reply:
So, essentially we wouldn't need any of these extra functions and could authorize our http request from the /render ?   

Current implementation of [new_request]( https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/run/markdown-preview/editor/render.py#L21)

Current implementation of [/render](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/run/markdown-preview/editor/main.py#L32)

Reply:
I like the simplification, since it is hard to communicate the local issues with the metadata server. For Java, the [request](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/run/markdown-preview/editor/src/main/java/com/example/cloudrun/RenderController.java#L64) would simplify to this [example](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/fca10f8f2e8d530577e9376608410669675fb7b4/run/authentication/src/main/java/com/example/cloudrun/Authentication.java#L36).

Reply:
Nice. I wonder if we'd want to make a point of having the snippet use the HTTP client built-in to the library, and have the tutorial use a popular HTTP client library except for minting the token? That allows us to keep token minting snippets as close to the IAP samples as possible, but still show a library many developers may want to use directly.

Reply:
I would like to show that if we can communicate it well. It would be nice for devs that just want to add a token to the request lib they are already using than have to reimplement with google libs.